_posts/2017-08-17-grpc-flatbuffers.md

```diff
@@ -0,0 +1,75 @@
+---
+layout: post
+title: Announcing out of the box support for gRPC in the Flatbuffers serializaton library.
+published: true
+permalink: blog/flatbuffers
+author: Wouter van Oortmerssen
+company: Google
+company-link: https://www.google.com
+---
+
+The recent release of Flatbuffers [version 1.7](https://github.com/google/flatbuffers/releases) introduced truly zero-copy support for gRPC out of the box. 
+
+[Flatbuffers](https://google.github.io/flatbuffers/) is a serialization library that allows you to access serialized data without first unpacking it or allocating any 
+additional data structures. It was originally designed for games and other resource constrained applications, but is now finding use
+within many teams at Google, and outside (eg: Facebook, Netflix).
+
+<!--more-->
+
+Flatbuffers enables maximum throughput by directly using gRPC's slice buffers with zero-copy for the common use cases. An incoming rpc may be processed directly from gRPCs internal buffers, and you constructing new message will write directly to internal buffers without intermediate steps.
+
+So far, this is supported in the C++ implementation of FlatBuffers. There is also an implementation in Go, which is not entirely zero copy, but still very low on allocation cost (see below). Other languages are in the works.
+
+
+### Example Usage
+
+Start with an .fbs schema (similar to .proto, if you are familiar with protocol buffers) that declares an RPC service:
+
+#### Use Flatbuffers as an IDL
+
+```
+table HelloReply {
+  message:string;
+}
+
+table HelloRequest {
+  name:string;
+}
+
+table ManyHellosRequest {
+  name:string;
+  num_greetings:int;
+}
+
+rpc_service Greeter {
+  SayHello(HelloRequest):HelloReply;
+  SayManyHellos(ManyHellosRequest):HelloReply (streaming: "server");
+}
+```
+
+This can be compiled with: `flatc --cpp --grpc example.fbs` to get generated C++ code, much like in protocol buffers.
```

Comment:
To generate C++ code from this, you run `flatc --cpp --grpc example.fbs`, much like in protocol buffers.